### PR TITLE
ENH: don't drop ImportError tracebacks for lazily-loaded optional dependencies

### PR DESCRIPTION
**Pull request by neutrinoceros**
equivalent to https://github.com/yt-project/yt/pull/4226
see https://github.com/yt-project/yt/issues/4224 for context

### CODE REVIEW

File: unyt/_on_demand_imports.py [hunk 1]
```diff
@@ -40,7 +59,7 @@ def __init_subclass__(cls):
 
     def __new__(cls):
         if cls is OnDemand:
-            raise TypeError("The OnDemand base class cannot be instantiated.")
```
Comment (jzuhone) [inline unyt/_on_demand_imports.py:43]: The old spelling is correct in American English, FWIW

Comment (neutrinoceros) [reply unyt/_on_demand_imports.py:43]: I thought it was. Well, it only got there because I copy-pasted the code from yt again.

**Comment by ngoldbaum** [comment]
This has a conflict. I'd also prefer it if you removed the types.

**Comment by neutrinoceros** [comment]
rebased without the conflict, and typing-free

**Pull request merged**